### PR TITLE
Clear PySpark gateway environment variables for OSS SparkML flavor dependency inference

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## Related Issues/PRs

#6905 

## What changes are proposed in this pull request?

Clear gateway env variables for OSS SparkML flavor dependency inference to not attempt to point to the driver SparkSession context from within a subprocess.

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
But let's ask user to verify this PR before merging.

File: mlflow/utils/_capture_modules.py
Comment (WeichenXu123) [inline mlflow/utils/_capture_modules.py:120]: Could you add a logging here to log all environment variables,
and let user rerun and provide the logged environment variables, which helps debugging :)

Comment (harupy) [reply mlflow/utils/_capture_modules.py:120]: Do we really need to log all environment variables?

File: mlflow/utils/requirements_utils.py
Comment (BenWilson2) [inline mlflow/utils/requirements_utils.py:267]: Seems like the original error with not resolving to minio instance is due to the subprocess not having access to the env config.

**Comment by BenWilson2** [comment]
Add a timeout to the subprocess

File: mlflow/utils/requirements_utils.py
Comment (WeichenXu123) [inline mlflow/utils/requirements_utils.py:269]: Might be too long ?
I think 120 seconds by default is long enough.

**Comment by WeichenXu123** [comment]
Does the latest PR pass user's code test ?

File: mlflow/utils/_capture_modules.py
Comment (WeichenXu123) [reply mlflow/utils/_capture_modules.py:120]: Just for debugging ... we should remove it before merging

Comment (BenWilson2) [reply mlflow/utils/_capture_modules.py:120]: added temporary logging with notes to remove

**Comment by BenWilson2** [comment]
@WeichenXu123 confirmed that the changes work in the user's environment and that the env variables are now passed correctly to support the use of minio.

File: mlflow/utils/_capture_modules.py
Comment (dbczumar) [inline mlflow/utils/_capture_modules.py:134]: What happens if this file doesn't exist?

File: mlflow/utils/_capture_modules.py
Comment (dbczumar) [inline mlflow/utils/_capture_modules.py:135]: Nit: Can we format this so that each `config()` call occurs on separate line?

File: mlflow/utils/requirements_utils.py
Comment (dbczumar) [inline mlflow/utils/requirements_utils.py:224]: To confirm, can a timer be canceled after it expires?

File: mlflow/utils/requirements_utils.py
Comment (dbczumar) [inline mlflow/utils/requirements_utils.py:269]: Can we add this to `mlflow.environment_variables` with docs and also prefix it with `MLFLOW_`?

File: mlflow/utils/requirements_utils.py
Comment (dbczumar) [inline mlflow/utils/requirements_utils.py:276]: What does this do? If it's necessary, can we add a small inline code comment explaining why?

**Review by dbczumar — COMMENTED** [review]
@BenWilson2 Awesome stuff! Just a few small comments.

File: mlflow/utils/_capture_modules.py
Comment (BenWilson2) [reply mlflow/utils/_capture_modules.py:134]: It becomes a 'best effort' resolve if the file isn't there. See the configuration Factory for initialization: https://logging.apache.org/log4j/2.x/manual/configuration.html#AutomaticConfiguration 

(if it's not set it'll use the default config)

File: mlflow/utils/_capture_modules.py
Comment (BenWilson2) [reply mlflow/utils/_capture_modules.py:135]: definitely!

File: mlflow/utils/requirements_utils.py
Comment (BenWilson2) [reply mlflow/utils/requirements_utils.py:224]: excellent catch :) Updating!

File: mlflow/utils/requirements_utils.py
Comment (BenWilson2) [reply mlflow/utils/requirements_utils.py:269]: definitely!

File: mlflow/utils/requirements_utils.py
Comment (BenWilson2) [reply mlflow/utils/requirements_utils.py:276]: added context relating to the original minio issue on kubernetes from the original issue ticket (and a link to the issue for future reference)

File: mlflow/environment_variables.py
Comment (dbczumar) [inline mlflow/environment_variables.py:169]: Can we make this a bit friendlier for users? I don't think users understand these details, including the subprocess

**Review by dbczumar — APPROVED** [review]
LGTM with one comment: https://github.com/mlflow/mlflow/pull/6912/files#r986164232. Thanks @BenWilson2 !

File: mlflow/environment_variables.py
Comment (BenWilson2) [reply mlflow/environment_variables.py]: good point! changed!

File: mlflow/environment_variables.py
Comment (dbczumar) [inline mlflow/environment_variables.py:168]:
```suggestion
#: Specifies the ``timeout_seconds`` for MLflow Model dependency inference operations.
```

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6912/files#r986300263 is addressed. Thanks @BenWilson2 !

**Pull request merged**
